src/core/column/cumcountngroup.h

```diff
@@ -0,0 +1,81 @@
+//------------------------------------------------------------------------------
+// Copyright 2022 H2O.ai
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#ifndef dt_COLUMN_CUMCOUNTNGROUP_h
+#define dt_COLUMN_CUMCOUNTNGROUP_h
+#include "column/virtual.h"
+#include "parallel/api.h"
+#include "stype.h"
+
+namespace dt {
+
+template<bool CUMCOUNT>
+class CUMCOUNTNGROUP_ColumnImpl : public Virtual_ColumnImpl {
+  private:
+    size_t nrows_;
+    bool reverse_;
+    Groupby gby_;
+
+  public:
+    CUMCOUNTNGROUP_ColumnImpl(size_t nrows, bool reverse, const Groupby& gby)
+      : Virtual_ColumnImpl(nrows, SType::INT64),
+        nrows_(nrows),
+        reverse_(reverse),
+        gby_(gby)
+      {}
+
+    ColumnImpl* clone() const override {
+      return new CUMCOUNTNGROUP_ColumnImpl(nrows_, reverse_, gby_);
+    }
+
+    size_t n_children() const noexcept override {
+      return 0;
+    }
+
+    const Column& child(size_t i) const override {
+      xassert(i == 0);  (void)i;
+    }
+
+    bool get_element(size_t i, int64_t* out) const override{
+      size_t i0, i1;
+      auto grp_size = gby_.size();
+      for (size_t j = 0; j < grp_size; ++j){
+         gby_.get_group(j, &i0, &i1);
+        if (i>=i0 & i<i1) {
+          if (CUMCOUNT){
+          *out = reverse_?i1-1-i:i-i0;
+          } else {
+            *out = reverse_?grp_size-1-j:j;
+          }
+          break;
+         }
+      }
+      return true;
+
+    }
```

Comment:
I guess we can implement a binary search here, since the offsets array is sorted.

Reply:
What advantage does a binary search offer here? If you don't mind could you explain more

Reply:
It will just be faster in identifying the correct group. Right not you just go from group 0 all the way until the needed group. With the binary search you start with the middle group and then move to left or right depending on where your element should be.

Look here: https://en.wikipedia.org/wiki/Binary_search_algorithm#Algorithm